compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -3689,12 +3689,10 @@ private boolean isAssignableToUnionType(BType source, BType target, Set<TypePair
                     }
                 }
                 // readonly can match to a union similar to any|error
-                if (sMember.tag == TypeTags.READONLY) {
+                if (sMember.tag == TypeTags.READONLY && isAssignable(symTable.anyAndReadonlyOrError, targetUnion)) {
```

Comment:
The following program should be a compile-time error. right?
```
public function main() {
    readonly readonlyVal = 1;
    error? errOrNil = <readonly>readonlyVal; 
}

```

Reply:
I believe so. It's currently giving a runtime error, I'll check.

Reply:
In a case like
```
readonly x = 2;
int? y = <readonly>x;
```
```
incompatible types: 'readonly' cannot be cast to '(int? & readonly)'
```
, the error is in casting and not assignment, since the casting target type is changed to `int? & readonly`
In the `error` case this cast is currently considered castable, therefore since the type is taken as `error? & readonly`, the assignment is also passing. Should this be a cast error or a assignment error in compiler-time?

Reply:
This seems to be a bug. Shall we create a separate issue? The assignment is what should result in an error.

Reply:
Please mention the issue here.

Reply:
Created in #36580